[WIP] Timespan support

jogibear9988 opened the pull request:
Fix #3993
Fix #4306
Fix #4308
Fix #2950


---


Open Questions/Tasks

- Should we change the default type in Mapping Schema for TimeSpan to int64. I've done it, but is this okay?
- Can I access the Column Type from my builder? Cause I want also to support for example the usage of a Bigint Column in Postgres. I've done it now like this way: var dt = ((SqlField)((SqlExpression)timeSpan).Parameters[0]).ColumnDescriptor.DataType; but sure this is not the correct way

viceroypenguin commented on Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:
```diff
@@ -501,7 +501,7 @@ private TransformInfo ConvertExpressionTransformer(Expression e)
 					}
 
-					if (ma.Member.DeclaringType == typeof(TimeSpan))
+					/*if (ma.Member.DeclaringType == typeof(TimeSpan))
```
I'm not sure I'm comfortable with take 55 lines of code and simply commenting them out. If these lines are not needed, they should be removed entirely. The whole point of source control is that the lines are still available in history if they are needed for reverting.

jogibear9988 replied:
It's not yet finished. I will remove them if it is. I only need answers to my questions, the whole pull still needs work, and I will remove  the commented out code.

I only now recoginzed, that a few others things break

viceroypenguin commented on Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:
```diff
@@ -557,11 +557,56 @@ private TransformInfo ConvertExpressionTransformer(Expression e)
+						&& Nullable.GetUnderlyingType(ex.Right.Type) == typeof(TimeSpan))
+					{
+						var method = MemberHelper.MethodOf(
+												() => Sql.DateAdd(Sql.DateParts.Millisecond, 0, DateTime.MinValue));
```
spacing is off here. this should only be one-tab to the right of `var`, not six.

jogibear9988 replied:
I will look after the formating if everything works and is okay

viceroypenguin commented on Source/LinqToDB/Mapping/MappingSchema.cs:
```diff
@@ -1330,7 +1330,7 @@ public DefaultMappingSchema() : base(new DefaultMappingSchemaInfo())
 				AddScalarType(typeof(DateTime),        DataType.DateTime2);
 				AddScalarType(typeof(DateTimeOffset),  DataType.DateTimeOffset);
-				AddScalarType(typeof(TimeSpan),        DataType.Time);
+				AddScalarType(typeof(TimeSpan),        DataType.Int64);
```
This is likely to break existing consumers. It should be done as a separate PR with support from the entire team, and it would not have my support.

jogibear9988 replied:
Don't know what will break. But clearly the Datatype "Time" wich is for a Time in Day is not the correct one

viceroypenguin replied:
You may not like it, but historically, `TimeSpan` was used for the sql `time` type, so it is accurate to the historical and technical reality.

jogibear9988 replied:
Yeah, for sure if we won't do this change, I can remove. But atm. I'll leave it. But whole calculation with timespans will not work with type, so everyone who will the correct functionality need than to apply a custom mapping.

jods4 replied:
There are two sides to this coin.

If you think "what should `TimeSpan` map to", then you're right `time` is a very bad choice.
If you think "what should `time` map to", then some people thought `TimeSpan` was the most fitting type in early .net.

In modern .net `time` would have a natural fit as `TimeOnly`.
For `TimeSpan`, SQL:2008 defines an `INTERVAL` type that is supported by many RDBMS including Oracle, MySQL, Postgre, MariaDB, Ingres. As such this would be a better default mapping (that would be overriden by providers not supporting `INTERVAL`, such as MSSQL).

In any case we need to be careful with backward compatibility. Isn't changing this default mapping is going to break existing consumers that have `time` in their DB?

jogibear9988 commented:
Now a little bit more info:

At first, the pull request is not yet done, the code needs to be cleaned up (remove commented, fix spaceing, ...)
Also I now only tried mssql, oracle and postgres, the other providers will follow if the code is okay, in the way it works)

The complexity results in, that the TimeSpan datatype has no coresponding type in most databases, so the Ticks will be used (100 nanoseconds) in a bigint field. But in postgres we have a interval type, wich is used for timespan. So if I calculate a datediff automaticaly it should also be a interval type, for this I added the "DatePart.Interval" to the datediff function.
If the database has no interval type, it will return the Ticks.

jogibear9988 commented:
/azp run test-postgresql

jogibear9988 commented:
/azp run test-all

jogibear9988 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,137 @@
+				var timeSpan = builder.GetExpression("timeSpan");
+
+				var dt = DataType.Undefined;
+				if (timeSpan is SqlField f)
```
Is this the correct way to check for the type of the Field in the Database?

MaceWindu replied:
use `timeSpan.GetExpressionType()` helper

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

Shane32 commented:
> * Should we change the default type in Mapping Schema for TimeSpan to int64. I've done it, but is this okay?

I'd agree with @viceroypenguin here: historically, `TimeSpan` has mapped to `time`, and could significantly affect existing codebases.  Now with .NET 6's introduction of the `TimeOnly` data type, it would make more sense to map `TimeOnly` to `time` and `TimeSpan` to `bigint`.  But I would not change the mapping at this time, as this would be a "silent" break in the code: it could potentially cause old code to compile against the new version with no design-time notifications, but yet break production queries.

If everyone agreed that mapping `TimeOnly` to `time` and `TimeSpan` to `bigint` would be a better default mapping long-term, then I might consider removing the default mapping of `TimeSpan` altogether for a couple major versions, and then reintroducing it linked to `bigint`.  In this way, users would hopefully upgrade, find that they must set the mapping manually (via global option or whatever), and later on when they upgrade again, the mapping will already be set in their code to use `time` rather than `bigint`, so the new default mapping will not break their code.

Another possibility would be to add a deprecation notice (perhaps via included .NET analyzer) to inform users that the `TimeSpan` mapping will change in the future, recommending users to switch to `TimeOnly`.

Another point of consideration is 'what is `SqlClient` and Entity Framework's default mapping for `TimeSpan`?'  If the rest of the .NET community assumes that `TimeSpan` gets mapped to `time`, then it's probably best that we at linq2db keep the same default, and use some type of runtime option or mapping configuration to select `bigint`, even if that makes more sense in 2023 than when the default was designed in 2008 for EF 3.5.

jogibear9988 commented:
At the moment I've switched back to "time" cause some tests break.
My Problem is, that I cannot map Timespan differently for some providers then.
In my test, I now mapped it to bigint, but for postgres I like to map it to Interval. This is only possible with this ugly code now: 
https://github.com/linq2db/linq2db/pull/3994/files#diff-71d51f038daa00c195480c91854cacefd7ece6c3e109b5b48d7c99fb3d9e77adR37

If we change it, for sure it this should be a major version change. But for now it's also okay for me if we leave it.

jods4 commented:
Can we not put the change behind an option? That's what we usually do for this kind of breaking change.

jogibear9988 commented:
we could. atm, i'm looking that all unit tests work again. found some wich don't and then I'll look.
Where are those options normaly?

jogibear9988 commented:
And what is for example with th expressions in the expressions.cs, they are in a static field, so how should we make them optional?
https://github.com/linq2db/linq2db/pull/3994/files#diff-2a6a9163861fbdc4015c6f32a3373c85469cb401552aea366b2d2c07ede63fb5

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

igor-tkachev commented:
We do not need to change default behavior. It should be possible to map to both TimeSpan and int64 depending on object property/field type.

jogibear9988 commented:
> We do not need to change default behavior. It should be possible to map to both TimeSpan and int64 depending on object property/field type.
but we also could have a timespan inside of a where wich is not mapped to field in the database.
And you calculate timespans by subtracting two dates, and there it does also matter if you compare against a c# timespan.

jods4 commented:
> At the moment I've switched back to "time" cause some tests break. My Problem is, that I cannot map Timespan differently for some providers then. In my test, I now mapped it to bigint, but for postgres I like to map it to Interval. This is only possible with this ugly code now: https://github.com/linq2db/linq2db/pull/3994/files#diff-71d51f038daa00c195480c91854cacefd7ece6c3e109b5b48d7c99fb3d9e77adR37
> 
> If we change it, for sure it this should be a major version change. But for now it's also okay for me if we leave it.

Maybe that's the biggest issue that needs to be fixed.
Redefining type mapping should be a relatively easy thing to do, both at the provider-level (because some provider have support for better types, e.g. lots of providers have `interval` for `TimeSpan`), as well as at the user-level (because some mappings can be argued upon; or you're working with a legacy db that has weird conventions. `TimeSpan` in MSSQL could be an int64, a string, a time if <24h, or a UDT).

If users can easily redefine some mappings at startup then we don't need new options, that'd be good enough (and more flexible).

jogibear9988 commented:
> > At the moment I've switched back to "time" cause some tests break. My Problem is, that I cannot map Timespan differently for some providers then. In my test, I now mapped it to bigint, but for postgres I like to map it to Interval. This is only possible with this ugly code now: https://github.com/linq2db/linq2db/pull/3994/files#diff-71d51f038daa00c195480c91854cacefd7ece6c3e109b5b48d7c99fb3d9e77adR37
> > If we change it, for sure it this should be a major version change. But for now it's also okay for me if we leave it.
> 
> Maybe that's the biggest issue that needs to be fixed. Redefining type mapping should be a relatively easy thing to do, both at the provider-level (because some provider have support for better types, e.g. lots of providers have `interval` for `TimeSpan`), as well as at the user-level (because some mappings can be argued upon; or you're working with a legacy db that has weird conventions. `TimeSpan` in MSSQL could be an int64, a string, a time if <24h, or a UDT).
> 
> If users can easily redefine some mappings at startup then we don't need new options, that'd be good enough (and more flexible).

a time is always the wrong type, cause a time is not a difference

igor-tkachev commented:
> > but we also could have a timespan inside of a where wich is not mapped to field in the database.
> > And you calculate timespans by subtracting two dates, and there it does also matter if you compare against a c# timespan.

It would be nice to have an example to better understand your point.

jods4 commented:
@jogibear9988 
> a time is always the wrong type, cause a time is not a difference ;-)

True but some people use `TimeSpan` to store a "time of day", as there was no `TimeOnly` in .net until recently.
With this perspective, the mapping is not that unreasonable.

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
So, I made now a config switch, wich switches to the new timespan mapping.
Problem is, in the unit tests, the MappingSchema Static constructor is called before my test has the chance to switch the setting to true. So how could I solve this?
Problem is, Testbase staticly initializes 2 Mapping schemas, testInitialization inits the TestNoopProvider wich also Initializes Mapping Schemas.
And my Setting needs to be set to true before the DefaultMappingSchema is created

jogibear9988 commented:
Found a way to do it, changed the timeSpan mapping later in the mapping schema

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
cann anyone have a look, there are many test failures, but they seem unrelated to my pull req.

System.Data.SqlClient.SqlException : Cannot open database "TestData" requested by the login. The login failed.
Login failed for user 'sa'.

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
There is now one failed test, but this has nothing to do with the changes.

    Tests.xUpdate.BulkCopyTests.KeepIdentity_SkipOnInsertFalse("Oracle.21.Managed",null,ProviderSpecific,0)
      Expected: 4
      But was:  23

But my new functionality is completely behind a flag and not tested with the tests, cause they flag needs to be set before the initialization of linq2db.

jogibear9988 commented:
But some of the functionality is tested now (DateDiff_Big used on newer SQLServer Versions, ...)

jogibear9988 commented:
If thisis done:
https://github.com/linq2db/linq2db/issues/4401 then it /the flag) could also be changed later

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented on Tests/Linq/DataProvider/SqlServerFunctionsTests.cs:
```diff
@@ -648,10 +648,13 @@ public void DateAddTest3([IncludeDataSources(TestProvName.AllSqlServer2008Plus)]
-			var result = db.Select(() => SqlFn.DateAdd(SqlFn.DateParts.Hour, 1, TimeSpan.FromHours(2)));
-			Console.WriteLine(result);
-			Assert.That(result, Is.EqualTo(new TimeSpan(3, 0, 0)));
+			if (!LinqToDB.Common.Configuration.MapTimeSpanToIntervalType)
```
this I forgot to remove, I will in the next commit

jogibear9988 commented on Tests/Linq/DataProvider/SqlServerFunctionsTests.cs:
```diff
@@ -684,10 +687,13 @@ public void DateDiffTest3([IncludeDataSources(TestProvName.AllSqlServer2008Plus)
-			var result = db.Select(() => SqlFn.DateDiff(SqlFn.DateParts.Hour, TimeSpan.FromHours(2), TimeSpan.FromHours(3)));
-			Console.WriteLine(result);
-			Assert.That(result, Is.EqualTo(1));
+			if (!LinqToDB.Common.Configuration.MapTimeSpanToIntervalType)
```
this I forgot to remove, it's not needed, I will in the next commit

jogibear9988 commented on Tests/Linq/DataProvider/SqlServerFunctionsTests.cs:
```diff
@@ -720,10 +726,13 @@ public void DateDiffBigTest3([IncludeDataSources(TestProvName.AllSqlServer2016Pl
-			var result = db.Select(() => SqlFn.DateDiffBig(SqlFn.DateParts.Hour, TimeSpan.FromHours(2), TimeSpan.FromHours(3)));
-			Console.WriteLine(result);
-			Assert.That(result, Is.EqualTo(1));
+			if (!LinqToDB.Common.Configuration.MapTimeSpanToIntervalType)
```
this I forgot to remove, it's not needed, I will in the next commit

igor-tkachev commented on Source/LinqToDB/Linq/Expressions.cs:
```diff
@@ -27,10 +27,43 @@ namespace LinqToDB.Linq
+		/// or
+		/// mappingSchema.AddScalarType(typeof(TimeSpan), DataType.Interval);
+		/// </summary>
+		public static void EnableTimeSpanMapping()
```
Configuration methods/properties should be declared in the `Common.Configuration` class. Move this method to `Common.Configuration`. Also the name and description are confusing.

jogibear9988 replied:
Will move

jogibear9988 replied:
Have you an Idea for a better Name?

jogibear9988 replied:
changed to "MapTimeSpanToIntervalType"

igor-tkachev reviewed (changes requested):
Am I the only one who thinks this is all wrong?

jogibear9988 commented:
> Am I the only one who thinks this is all wrong?

Then tell me what is wrong? 
Problem is, I want to use DateDifferences in my Database, wich are compared against timespans, and this does not work. Cause it only works in some special cases in Linq2Db, as you see there is special handling in ExpressionBuilder.SqlBuilder.
The main Problem also is, that TimeSpans are mapped to DateBase type "Time" wich is not a timespan, but a Time of day.

I'll change this to an intervaltype on database who support (Postgres atm, maybe more but I've only worked on Postgres/Orcale/Sqlite/Mssql), and to an BigInt on others. And then I created a Function wich creates Interval/Bigint out of the TimeSpan.
I will add more Databases, if the current code is reviewed (and maybe merged). Cause I can test the others only on the Build Pipeline here, as I. have not installed them.

igor-tkachev commented:
> Then tell me what is wrong?

This doesn't look like a solution to the whole problem, it looks like a quick temporary fix for a very specific case. If you run all the tests with your fix, many tests will fail. That's why I say that this is a fix for a special case. I think it's better if you provide tests that do not work for your scenario and we will try to find more appropriate solution.

jogibear9988 commented:
> > Then tell me what is wrong?
> 
> This doesn't look like a solution to the whole problem, it looks like a quick temporary fix for a very specific case. If you run all the tests with your fix, many tests will fail. That's why I say that this is a fix for a special case. I think it's better if you provide tests that do not work for your scenario and we will try to find more appropriate solution.

The tests fail, if a TimeSpan is mapped to "time", wich is wrong. The rest of the tests work. How should I fix it, without breaking the old behavior. 
What on my code looks wrong? For me the old special cases look wrong. 
Mine supports much more cases, look at my new test. I'd like to make mine the standard, and set a flag on the tests wich do not work (cause they map to a wrong database type), but the main complain here was, to not do a breaking change.

jogibear9988 commented:
In Exprpressionbuilder.Sqlbuilder this is the old code:
(wich I also keep, cause I need it when the mapping is wrong

						if (ma.Member.DeclaringType == typeof(TimeSpan))
						{
							switch (ma.Expression!.NodeType)
							{
								case ExpressionType.Subtract:
								case ExpressionType.SubtractChecked:

									Sql.DateParts datePart;

									switch (ma.Member.Name)
									{
										case "TotalMilliseconds": datePart = Sql.DateParts.Millisecond; break;
										case "TotalSeconds": datePart = Sql.DateParts.Second; break;
										case "TotalMinutes": datePart = Sql.DateParts.Minute; break;
										case "TotalHours": datePart = Sql.DateParts.Hour; break;
										case "TotalDays": datePart = Sql.DateParts.Day; break;
										default: return new TransformInfo(e);
									}

									var ex = (BinaryExpression)ma.Expression;
									if (ex.Left.Type == typeof(DateTime)
										&& ex.Right.Type == typeof(DateTime))
									{
										var method = MemberHelper.MethodOf(
												() => Sql.DateDiff(Sql.DateParts.Day, DateTime.MinValue, DateTime.MinValue));

										var call   =
												Expression.Convert(
													Expression.Call(
														null,
														method,
														Expression.Constant(datePart),
														Expression.Convert(ex.Right, typeof(DateTime?)),
														Expression.Convert(ex.Left,  typeof(DateTime?))),
													typeof(double));

										return new TransformInfo(ConvertExpression(call));
									}
									else
									{
										var method = MemberHelper.MethodOf(
												() => Sql.DateDiff(Sql.DateParts.Day, DateTimeOffset.MinValue, DateTimeOffset.MinValue));

										var call   =
												Expression.Convert(
													Expression.Call(
														null,
														method,
														Expression.Constant(datePart),
														Expression.Convert(ex.Right, typeof(DateTimeOffset?)),
														Expression.Convert(ex.Left,  typeof(DateTimeOffset?))),
													typeof(double));

										return new TransformInfo(ConvertExpression(call));
									}
							}
						}
					}

and this is my new code. It nearly looks the same, but it supports more cases (DateTime & DateTime, DateTime & TimeSpan, DatetimeOffset & DateTimeOffset, DateTimeOffset and TimeSpan).


					var ex = (BinaryExpression)e;
					var leftType = (Nullable.GetUnderlyingType(ex.Left.Type) ?? ex.Left.Type);
					var righttype = (Nullable.GetUnderlyingType(ex.Right.Type) ?? ex.Right.Type);
					
					if (leftType == typeof(DateTime) && righttype == typeof(TimeSpan))
					{
						var method = MemberHelper.MethodOf(
												() => Sql.DateAdd(Sql.DateParts.Millisecond, 0, DateTime.MinValue));

						Expression value = Expression.Convert(Expression.Property(Expression.Convert(ex.Right,  typeof(TimeSpan)), "TotalMilliseconds"), typeof(double?));
						
						if (ex.NodeType == ExpressionType.Subtract || ex.NodeType == ExpressionType.SubtractChecked)
							value = Expression.Convert(Expression.Multiply(Expression.Convert(value, typeof(double)), Expression.Constant((double)-1)), typeof(double?));

						var call   = Expression.Convert(
													Expression.Call(
														null,
														method,
														Expression.Constant(Sql.DateParts.Millisecond),
														value,
														Expression.Convert(ex.Left, typeof(DateTime?))),
													 typeof(DateTime?));

						return new TransformInfo(ConvertExpression(call));
					}

					if (leftType == typeof(DateTimeOffset) && righttype == typeof(TimeSpan))
					{
						var method = MemberHelper.MethodOf(
												() => Sql.DateAdd(Sql.DateParts.Millisecond, 0, DateTimeOffset.MinValue));

						Expression value = Expression.Convert(Expression.Property(Expression.Convert(ex.Right,  typeof(TimeSpan)), "TotalMilliseconds"), typeof(double?));

						if (ex.NodeType == ExpressionType.Subtract || ex.NodeType == ExpressionType.SubtractChecked)
							value = Expression.Convert(Expression.Multiply(Expression.Convert(value, typeof(double)), Expression.Constant((double)-1)), typeof(double?));

						var call   = Expression.Convert(
													Expression.Call(
														null,
														method,
														Expression.Constant(Sql.DateParts.Millisecond),
														value,
														Expression.Convert(ex.Left,  typeof(DateTimeOffset?))),
													 typeof(DateTimeOffset?));

						return new TransformInfo(ConvertExpression(call));
					}

					if (leftType == typeof(DateTime) && righttype == typeof(DateTime))
					{
						var method = MemberHelper.MethodOf(
												() => Sql.DateDiffInterval(DateTime.MinValue, DateTime.MinValue));
						var call   =			Expression.Convert(
													Expression.Call(
														null,
														method,
														Expression.Convert(ex.Right, typeof(DateTime?)),
														Expression.Convert(ex.Left,  typeof(DateTime?))),
													typeof(TimeSpan));

						return new TransformInfo(ConvertExpression(call));
					}

					if (leftType == typeof(DateTimeOffset) && righttype == typeof(DateTimeOffset))
					{
						var method = MemberHelper.MethodOf(
												() => Sql.DateDiffInterval(DateTimeOffset.MinValue, DateTimeOffset.MinValue));
						var call   =            Expression.Convert(
													Expression.Call(
														null,
														method,
														Expression.Convert(ex.Right, typeof(DateTimeOffset?)),
														Expression.Convert(ex.Left,  typeof(DateTimeOffset?))),
													typeof(TimeSpan));

						return new TransformInfo(ConvertExpression(call));
					}

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
@igor-tkachev 

I enabled now my change for all tests, 902 of 798236 fail.

SqlServe2008, Access => Don't work cause of overflow, I've not yet looked into this, at first I wanted to get an okay that it would work so.

SqlServer => bigint is incompatible with time, there is nothing I could do. Wen need to change the test if we want mapping to int64 is the standard, or leave it as is, and use my flag.

For example (from the test)

      Assert.That(TestType<TimeSpan?>      (conn, "timeDataType",           DataType.Time,           "AllTypes2"), Is.EqualTo(new TimeSpan(0, 12, 12, 12, 12)));

so this will not work when I change the type, but it is needed for correct timespan support. Cause of this I added the configuration flag.

So what is wrong with my code? it only breaks tests wich specially depend on TimeSpan mapped to Time (cause of this my change is optional) or breaks some tests on old database (cause I've not yet looked after, but will if it's okay as I've done it)

I don't understand what should be wrong, or what should I do different. At the moment, if you use some calculations with DateTime & TimeSpan and want to do that in database, it will not work with linq2db without my patch.

MaceWindu commented on Source/LinqToDB/Common/Configuration.cs:
```diff
@@ -52,6 +52,42 @@ public static Delegate CompileExpression(this LambdaExpression expression)
+		{
+			MappedTimeSpanToIntervalType = true;
+
+			Expressions.MapMember<TimeSpan>(p => p.Days, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.Days, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.TotalDays, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.TotalDays, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.Hours, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.Hours, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.TotalHours, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.TotalHours, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.Minutes, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.Minutes, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.TotalMinutes, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.TotalMinutes, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.Seconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.Seconds, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.TotalSeconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.TotalSeconds, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.Milliseconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.Milliseconds, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.TotalMilliseconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.TotalMilliseconds, p)!.Value));
+
+#if NET7_0_OR_GREATER
+			Expressions.MapMember<TimeSpan>(p => p.Microseconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.Microseconds, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.TotalMicroseconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.TotalMicroseconds, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.Nanoseconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.Nanoseconds, p)!.Value));
+			Expressions.MapMember<TimeSpan>(p => p.TotalNanoseconds, (Expression<Func<TimeSpan, long>>)(p => LinqToDB.Sql.TimeSpanPart(LinqToDB.Sql.TimeSpanParts.TotalNanoseconds, p)!.Value));
+#endif
```
this should go to Expressions.cs

igor-tkachev replied:
Configuration method should here, but it's implementation can be moved.

jogibear9988 replied:
I had it there, I moved it cause of comments.
=> I'll move it back

MaceWindu commented on Source/LinqToDB/Common/Configuration.cs:
```diff
@@ -52,6 +52,42 @@ public static Delegate CompileExpression(this LambdaExpression expression)
+		/// <summary>
+		/// Fixes wrong TimespanMapping and uses a Interval Type if the Database has one (if not it uses Int64)
+		/// </summary>
+		internal static bool MappedTimeSpanToIntervalType;
```
this is not how client<->db type mapping works. We don't use options for that but:
- to specify database type we use `DataType` enumeration
- to change default mapping `MappingSchema.SetDataType`  method used

jogibear9988 replied:
I need a setting cause at the moment TimeSpan is Mapped to a wrong database Type. So I need to know if you would map to correct one so other Code in ExpressionBuilder.SqlBuilder.cs is used

jogibear9988 replied:
I removed now the setting, I look in the expression.sqlbuilder for:

			var timeSpanDatatype = this.MappingSchema.GetDataType(typeof(TimeSpan));
			var timeSpanHasIntervaltype = timeSpanDatatype.Type.DataType == DataType.Interval || timeSpanDatatype.Type.DataType == DataType.Int64;

MaceWindu commented on Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:
```diff
@@ -501,33 +501,35 @@ private TransformInfo ConvertExpressionTransformer(Expression e)
+					if (!Configuration.MappedTimeSpanToIntervalType)
 					{
-						switch (ma.Expression!.NodeType)
+						if (ma.Member.DeclaringType == typeof(TimeSpan))
```
this code needs test converage with mappings enabled

jogibear9988 replied:
I don't know what you mean. If this setting is enabled, this code is not used. Normaly it is disabled. I only enabled it as default teporally to see wich tests will fail.

MaceWindu commented on Source/LinqToDB/Mapping/MappingSchema.cs:
```diff
@@ -1378,7 +1383,9 @@ public DefaultMappingSchema() : base(new DefaultMappingSchemaInfo())
 				AddScalarType(typeof(DateTimeOffset),  DataType.DateTimeOffset);
-				AddScalarType(typeof(TimeSpan),        DataType.Time);
+
+				AddScalarType(typeof(TimeSpan), DataType.Int64);
```
no changes to default mappings

jogibear9988 replied:
This is only in a temporary commit, to see wich tests fail when enabled, will remove it again

MaceWindu commented on Source/LinqToDB/Mapping/MappingSchema.cs:
```diff
@@ -36,6 +36,11 @@ namespace LinqToDB.Mapping
 	{
+		static MappingSchema()
+		{
+			Configuration.MapTimeSpanToIntervalType();
```
???

jogibear9988 replied:
This is only in a temporary commit, to see wich tests fail when enabled, will remove it again

MaceWindu commented on Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:
```diff
@@ -554,14 +556,103 @@ private TransformInfo ConvertExpressionTransformer(Expression e)
+				case ExpressionType.Add:
+				case ExpressionType.AddChecked:
+				case ExpressionType.Subtract:
+				case ExpressionType.SubtractChecked:
```
there is no need to add code like that. TimeSpanPart builder should handle both db types

jogibear9988 replied:
Time Span Part Build handles these cases. But he needs to be call in all of them. 
It's not different from the already existing code...

jogibear9988 replied:
@MaceWindu 

how can I do this? How can the build be called when having a Add/Subtract expression?

I tried to Map like this:

        Expressions.MapMember<DateTime, DateTime, TimeSpan>((p1, p2) => p1 - p2, (Expression<Func<DateTime, DateTime, TimeSpan>>)((p1, p2) => LinqToDB.Sql.DateDiffInterval(p2, p1)!.Value));
        Expressions.MapMember<DateTimeOffset, DateTimeOffset, TimeSpan>((p1, p2) => p1 - p2, (Expression<Func<DateTimeOffset, DateTimeOffset, TimeSpan>>)((p1, p2) => LinqToDB.Sql.DateDiffInterval(p2, p1)!.Value));
        Expressions.MapMember<DateTime, TimeSpan, DateTime>((p1, p2) => p1 + p2, (Expression<Func<DateTime, TimeSpan, DateTime>>)((p1, p2) => LinqToDB.Sql.DateAdd(p1, p2)!.Value));
        Expressions.MapMember<DateTimeOffset, TimeSpan, DateTimeOffset>((p1, p2) => p1 + p2, (Expression<Func<DateTimeOffset, TimeSpan, DateTimeOffset>>)((p1, p2) => LinqToDB.Sql.DateAdd(p1, p2)!.Value));
		
but it did not work, causes a exception.

MaceWindu replied:
We already have code to handle those operations above. Just update that code if it miss something instead of adding duplicate branch

jogibear9988 replied:
No we do not yet support Adding a TimeSpan to a Date.

But I found it, I can add a Expression with MapBinary

MaceWindu commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,170 @@
+			{
+				var part     = builder.GetValue<TimeSpanParts>("part");
+				var partStr  = TimeSpanPartToStr(part);
+				var timeSpan = builder.GetExpression("timeSpan");
```
should inspect DataType of timeSpan expression and generate corresponding SQL

MaceWindu commented on Source/LinqToDB/Sql/Sql.Types.cs:
```diff
@@ -125,6 +125,10 @@ public static class Types
 
+			[Property(PN.PostgreSQL, "Interval", ServerSideOnly = true, CanBeNull = false)]
+			[Property("BigInt", ServerSideOnly = true, CanBeNull = false)]
+			public static TimeSpan TimeSpan => TimeSpan.Zero;
```
Is it used? If no I prefer to avoid new "mappings" in this class, especially taking into account that it doesn't workproperly with type mapping

jogibear9988 replied:
I've removed it

MaceWindu reviewed (approved):
for starters

MaceWindu reviewed (changes requested):
1

MaceWindu commented:
Also it should contain test for #2950 here

jogibear9988 commented:
Oracle does not yet work after my refactoring, also my Tests are not yet all workin.
But I'll look after this.

@MaceWindu 
Could you take a look again, if I could implement functionality like this now.
I removed all my Code from the SqlBuilder and added a DateAdd for Date+TimeSpan

It's not yet finished, need to look after oracle and other databases, and create better tests (wich also test the results).

But I want to know if the way I'm going is okay.

And one question, how can I force a expression executed on the client side, to be executed on the server?

jogibear9988 commented:
I now added a way to Map custom Unary Expressions:

      MapUnary<TimeSpan, TimeSpan?>((TimeSpan t) => - t, (t) => Sql.NegateInterval(t)!.Value);
      MapUnary<TimeSpan?, TimeSpan?>((TimeSpan? t) => -t, (t) => Sql.NegateInterval(t)!.Value);

jogibear9988 commented:
I enabled now all databases, to see wich ones I need to fix

jogibear9988 commented:
Not supported:

        SqlServer < 2016 (cause of Precission of DateDiff) and SQL CE
        Access (cause of no support for BigInt, and Timespan mapped to Ticks)

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
Does anyone know, wich databases have a native interval type?

  - Postgres
  - Oracle

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jods4 commented on Source/LinqToDB/Linq/Builder/ExpressionBuilder.QueryBuilder.cs:
```diff
@@ -765,7 +765,27 @@ bool PreferServerSide(Expression expr, bool enforceServerSide)
+							if (l != null)
+							{
+								var body = l.Body.Unwrap();
+								var newExpr = body.Transform((l, unary), static (context, wpi) =>
```
Instead of copy-pasting this somewhat obscure code from BinaryExpression, can we extract a function?
Suggestion:
```cs
static Expression InlineLambdaParameters(LambdaExpression lambda, Expression[] values)
{
  return lambda
    .Body.Unwrap()
    .Transform((parameters: lambda.Parameters, values), static (context, visited) => 
    {
      if (visited.NodeType != ExpressionType.Parameter)
        return visited;
      int i = Array.IndexOf(context.Parameters, visited);
      return i < 0 ? visited : context.values[i];
    });
}
```
I wouldn't be surprised if a similar function already exists somewhere in linq2db codebase.

jogibear9988 replied:
how should this work? 
in binary I need to return context.binary.Left or return context.binary.right;

jods4 commented on Source/LinqToDB/Linq/Expressions.cs:
```diff
@@ -125,6 +125,97 @@ public static void MapMember<T>(string providerName, Expression<Func<T,object?>>
+
+		static UnaryExpression GetUnaryNode(Expression expr)
+		{
+			while (expr.NodeType == ExpressionType.Convert || expr.NodeType == ExpressionType.ConvertChecked || expr.NodeType == ExpressionType.TypeAs)
```
Can we use latest C# compiler?
Let's shorten this long lines: `expr.NodeType is ExpressionType.Convert or ExpressionType.ConvertChecked or ExpressionType.TypeAs`

jogibear9988 replied:
done

jods4 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,471 @@
+				TimeSpanParts.Ticks             => (long)timeSpan.Value.Ticks,
+				_ => throw new InvalidOperationException(),
+			};
+			throw new NotImplementedException();
```
Unreachable code?

jods4 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,471 @@
+		[Extension(PN.SQLite,     "",																													        ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(TimeSpanPartBuilderSqlite))]
+		[Extension(PN.Oracle,     "(extract(second from {0}) + extract(minute from {0}) * 60 + extract(hour from {0}) * 3600 + extract(day from {0}) * 86400)", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(TimeSpanPartBuilderIntervalType))]
+		[Extension(PN.PostgreSQL, "(extract(second from {0}) + extract(minute from {0}) * 60 + extract(hour from {0}) * 3600 + extract(day from {0}) * 86400)", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(TimeSpanPartBuilderIntervalType))]
+		public static long? TimeSpanPart([SqlQueryDependent] TimeSpanParts part, [ExprParameter] TimeSpan? timeSpan)
```
Add `[NotNullIfNotNull]` on the returned value

jogibear9988 replied:
done

jods4 commented on Source/LinqToDB/Sql/Sql.DateTime.cs:
```diff
@@ -1118,10 +1234,230 @@ public void Build(ISqExtensionBuilder builder)
+		[Extension(PN.Access,       "",              BuilderType = typeof(DateDiffIntervalBuilderAccess))]
+		[Extension(PN.ClickHouse,   "",              BuilderType = typeof(DateDiffIntervalBuilderClickHouse))]
+		/* Returns the Native Database Interval type, or the Timespan Ticks (100ns) */		
+		public static TimeSpan? DateDiffInterval(DateTime? startDate, DateTime? endDate)
```
Why a static `Sql.DateDiffInterval(d1, d2)` rather than compiling `operator -`? 
`d1 - d2` is valid C# returning a `TimeStamp`.

Moreover the operator handles nullability properly, this method would need some annotations for proper nullability flow (return type is not null when both operands are not null: is this even doable in C#?)

We should also handle `DateTimeOffset`; and `TimeOnly` would be nice, too.

In my opinion `DateOnly` does not make sense, neither do combinations of `DateTimeOffset` x `DateTime` x `TimeOnly`.

jods4 replied:
After reviewing more code, I think having `Sql.DateDiffInterval` is a good addition to `d1 - d2` if we add a `unit` enum that lets user pick the actual `datediff` unit. This works-around all the precisions problems I mentionned above, while keeping a nice 1:1 relationship between C# code and SQL.

For `operator -` no precision can be specified so it'd be nice to pick a somewhat usable default.
In a comment above I proposed `microsecond` where 64 bits numbers are used, and `second` as a compromise for 32 bits numbers.

jogibear9988 replied:
> After reviewing more code, I think having `Sql.DateDiffInterval` is a good addition to `d1 - d2` if we add a `unit` enum that lets user pick the actual `datediff` unit. This works-around all the precisions problems I mentionned above, while keeping a nice 1:1 relationship between C# code and SQL.
> 
> For `operator -` no precision can be specified so it'd be nice to pick a somewhat usable default. In a comment above I proposed `microsecond` where 64 bits numbers are used, and `second` as a compromise for 32 bits numbers.
> 
> EDIT: basically `DateDiffInterval(DateParts, DateTime, DateTime)` is consistent with existing functions above.

I think I need to use 100nanoseconds, cause it could be compared to a timespan. And that is always encoded as ticks (changeing it would be a breaking change, and also increasing complexity)

jogibear9988 replied:
This is new, but timespans are always suppored to use in a parameter, and they are serialized to sql as their ticks value

jods4 replied:
Sorry I'm blunt, I still don't see the specific case where it happens.
Can you give an example of a line of code that works today and would be broken?

jogibear9988 replied:
what about this code:

     aaa.where(x=>x.DateTime1-x.DateTime2 < TimeSpan.FromSeconds(400))

TimeSpan is always converted to 100nanoseconds (not through my code this is existing behavior) in the sql parameter, but the datediffinterval would not return nannoseocnds in your case

jogibear9988 replied:
i now changed to use milliseconds in my calls, but this cause also overflow in sqlserver

jods4 commented on Source/LinqToDB/Sql/Sql.DateTime.cs:
```diff
@@ -1118,10 +1234,230 @@ public void Build(ISqExtensionBuilder builder)
+				var startDate = builder.GetExpression(0);
+				var endDate = builder.GetExpression(1);
+				var expStr = "(CAST ({1} as TIMESTAMP) - CAST ({0} as TIMESTAMP))";
+				builder.ResultExpression = new SqlExpression(typeof(int), expStr, startDate, endDate);
```
Why is this typed as `int` when it's a SQL `INTERVAL`, which would be a `TimeSpan` in C#?
It would be nice to have a comment explaining.

jogibear9988 replied:
maybe because it's an error

jods4 commented on Source/LinqToDB/Sql/Sql.DateTime.cs:
```diff
@@ -1118,10 +1234,230 @@ public void Build(ISqExtensionBuilder builder)
+				var startdate  = builder.GetExpression(0);
+				var endDate    = builder.GetExpression(1);
+
+				builder.ResultExpression = new SqlFunction(typeof(int), "Nano100_Between", startdate, endDate); ;
```
Extra semi-colon at end

jods4 commented on Source/LinqToDB/Sql/Sql.DateTime.cs:
```diff
@@ -1118,10 +1234,230 @@ public void Build(ISqExtensionBuilder builder)
+				var startdate = builder.GetExpression(0);
+				var endDate   = builder.GetExpression(1);
+
+				builder.ResultExpression = new SqlExpression(typeof(int), builder.Expression + "(nanosecond, {0}, {1}) / 100", startdate, endDate);
```
Even though C# TimeSpan has a precision of 100ns, most RDBMS do not support that and I believe we need better defaults for usability.

Consider for example SQL SERVER.
Before MSSQL 2016, only `datediff` is supported and it would overflow at a minute already, making this code totally unusable (even at the `second` precision, an interval of 70 years cannot be computed).
Since MSSQL 2016, `datediff_big` is available and would never overflow except at the `nanosecond` precision, where an interval of 293 years or more fails. Depending on your field of work, you could be more interested in years than nanosecond.

In many case (but not all) a lot of RDMBS types do not support ns precision at all. 

If we want to go "simple", I'd suggest:
- using `microsecond` precision where big ints are supported as it would never fail (and TimeSpan only has a 0.1us precision anyway, not even close to 1ns).
- using `second` where only 32 bits precision is supported -- this is a limitation in both small (<1s) and large (>70y) intervals.

If someone needs a different precision (big or small), they could create a custom `[Expression]` static function for that purpose.

jogibear9988 replied:
i now use milliseconds for the add call

jods4 commented on Source/LinqToDB/Sql/Sql.DateTime.cs:
```diff
@@ -487,12 +502,51 @@ sealed class DateAddBuilder : IExtensionCallBuilder
+									new SqlExpression("millisecond", Precedence.Primary), builder.Div(new SqlBinaryExpression(typeof(long), number, "%", new SqlValue(3600_000_000_0), Precedence.Multiplicative), 1000_0),
+										new SqlFunction(typeof(DateTime?), builder.Expression,
+											new SqlExpression("nanosecond", Precedence.Primary), new SqlBinaryExpression(typeof(long), new SqlBinaryExpression(typeof(long), number, "%", new SqlValue(3600_000_000_0), Precedence.Multiplicative), "%", new SqlValue(1000_0), Precedence.Multiplicative), date)));
+				}
```
I really don't like this part, let's see what other team member think about it.
As a user I'd hate to find out that my simple C# `DateAdd(DateParts.Nanosecond, 10, x)` turned into a triple nested for possibly no reasons, as a lot of rdbms types would either work or not support that kind of precisions anyway.

What makes things worse is that the `number` expression, which is repeated 3 times could be anything including complex subqueries.

jogibear9988 replied:
okay, will remove

jods4 commented on Source/LinqToDB/Sql/Sql.DateTimeOffset.cs:
```diff
@@ -152,9 +164,77 @@ public void Build(ISqExtensionBuilder builder)
+		[Extension(PN.PostgreSQL,   "",              BuilderType = typeof(DateDiffIntervalBuilderPostgreSql))]
+		[Extension(PN.Access,       "",              BuilderType = typeof(DateDiffIntervalBuilderAccess))]
+		[Extension(PN.ClickHouse,   "",              BuilderType = typeof(DateDiffIntervalBuilderClickHouse))]
+		public static TimeSpan? DateDiffInterval(DateTimeOffset? startDate, DateTimeOffset? endDate)
```
Basically same comments as on the Datetime version:
- I would love an explicit DateParts unit;
- and a reasonable default unit based on platform for the `-` operator.

Also would like a non-nullable operator because otherwise we have to do annoying `.Value` in query even when columns are non-nullable (as TimeSpan is a struct).

jods4 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,471 @@
+		[Extension(PN.ClickHouse, "", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(DateTimeAddIntervalBuilderNanoseconds))]
+		[Extension(PN.SapHana,    "", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(DateTimeAddIntervalBuilderMilliseconds))]
+		[Extension(PN.Sybase,     "", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(DateTimeAddIntervalBuilderMicroseconds))]
+		public static DateTime? DateAdd(DateTime? date, TimeSpan? timeSpan)
```
Please add support for non-nullable variations

jods4 replied:
In fact, non-nullable support would come for free if we compiled `DateTime + TimeSpan` operator, why not?
`Sql.DateAdd(unit, number, date)` has the added benefit of accepting a `unit` and mapping to the same-named SQL function.

This `DateAdd` overload does not, so I don't see the point over `operator +`.

jogibear9988 replied:
I don't get what you mean here...
I've not added a non nun, cause for DateAdd there was also not a non null version.
Linq at all is supported with nullable and non nullable

jods4 replied:
Ok so after reviewing everything I notice this is the mapped expression for `operator +`.
Let's make it private and we benefit from operators being lifted so it works for combination of nullables and has proper nullability data flow.

What I meant is that not having non-nullable overloads forces me to add `!.Value` when I use this function with non-nullable dates.

For example:
```cs
DateTime dt; // non-null
TimeSpan ts; // non-null

// What I intuitively want (does not compile):
var result = Sql.DateAdd(dt, ts);
// What I must write:
var result = Sql.DateAdd(dt, ts)!.Value;
```

This is easily fixed by adding overloads, but if we hide this static function and support only `dt + ts` there's no point here.

jogibear9988 replied:
made it internal

jods4 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,471 @@
+		[Extension(PN.ClickHouse, "", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(DateTimeAddIntervalBuilderNanoseconds))]
+		[Extension(PN.SapHana,    "", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(DateTimeAddIntervalBuilderMilliseconds))]
+		[Extension(PN.Sybase,     "", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(DateTimeAddIntervalBuilderMicroseconds))]
+		public static DateTimeOffset? DateAdd(DateTimeOffset? date, TimeSpan? timeSpan)
```
Same: non-nullable variations

jods4 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,471 @@
+		}
+
+		[Extension("", ServerSideOnly = false, PreferServerSide = false, BuilderType = typeof(NegateIntervalBuilder))]
+		public static TimeSpan? NegateInterval(TimeSpan? timeSpan)
```
Why not support unary operator `TimSpan` unary `-` operator?

jods4 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,471 @@
+
+				if (dt == DataType.Int64)
+				{
+					var p = Expression.Call(
```
I'd like to suggest a different translation for Oracle: you can add a number to date, it's understood as a number of days
```sql
date + expr / TicksPerDay
```

jogibear9988 replied:
i don't like this. Whats the precision? Current code works, so why the change?

jods4 commented on Source/LinqToDB/Sql/Sql.TimeSpan.cs:
```diff
@@ -0,0 +1,471 @@
+				var p = Expression.Call(
+						null,
+						MethodHelper.GetMethodInfo(TimeSpanPart, TimeSpanParts.TotalNanoseconds, (TimeSpan?)TimeSpan.Zero),
+						Expression.Constant(TimeSpanParts.TotalNanoseconds),
```
I'll make the same remark as in `Sql.DateTime`: using nanosecond precisions by default is very unusable.
In MSSQL for example (all releases) the doc says "an expression that can resolve to an `int`" (I did not test).
If the doc is accurate, this means 32 bits, which is an incredibly small interval for nanoseconds.

I would pick a different default unit for `date + timespan`, probably `second`, which gives IMHO the most usable range.

As this applies to users of RDBMS without proper interval support, if they need a different precision they can always convert their expression to a long (I believe `Sql.NoConvert<long>(fake_timespan_column)` does the trick?) and pass it to a manually crafted `Sql.DateAdd()`.

jogibear9988 replied:
chnaged to milliseconds

jods4 commented on Source/LinqToDB/Linq/Builder/ExpressionTreeOptimizationContext.cs:
```diff
@@ -935,6 +935,26 @@ public bool PreferServerSide(Expression expr, bool enforceServerSide)
+							if (l != null)
+							{
+								var body = l.Body.Unwrap();
+								var newExpr = body.Transform((l, unary), static (context, wpi) =>
```
This is the same copy-pasted code that I suggested we factor out in ExpressionBuilder.QueryBuilder.cs.
Would be nice to reuse the same function here (+ binary version below).

jogibear9988 replied:
not possible, se other comment

jods4 commented on Source/LinqToDB/Linq/Expressions.cs:
```diff
@@ -383,6 +474,41 @@ public static void SetGenericInfoProvider(Type type)
+			Dictionary<Tuple<ExpressionType,Type>,IExpressionInfo>? dic;
+
+			var unaries = _unaries.Value;
+			var key      = Tuple.Create(unaryExpression.NodeType, unaryExpression.Operand.Type);
```
@all Can we add `System.ValueTuple` as an additional package reference for net45 and net46, and use `ValueTuple` internally for the reduced allocation and convenient C# syntax?

jods4 commented on Source/LinqToDB/Linq/Expressions.cs:
```diff
@@ -1702,6 +1831,70 @@ public void SetInfo(MappingSchema mappingSchema)
+		public static void AddTimeSpanMappings()
+		{
+			Configuration.DisableLegacySqlBuilderDateDiffCalls = true;
+
```
I like this approach 👍
We can make it more performant but I'll wait until @sdanyliv completes his refactoring, some ideas:
- `MappingSchema` should carry a provider-specific dictionary that merges all applicable mappings, so that runtime lookup is faster.
- In the spirit of building this merged dictionary, I would move provider-specific mappings into provider code (and remove the strings). This is cleaner but also faster at startup as we don't need create all expressions specific to providers we don't use.
- It'd be nice for `Map*` to support a builder as second parameter. The current code works but creates many intermediary objects and transformations that slow down compilation.
- In anticipation of that 3rd point, it'd be nice to not make public any static `Sql.*` function that could be removed in the future if we reference a builder directly.

I'll tackle these when @sdanyliv refactoring is merged.

jogibear9988 commented:
I now added an extra overload for NegateInterval for sqlserver.

cause if I add a timespan to datetime, sqlserver only supports int, so it overflows very early. I now split the call in two add operations, wich result in following sql:

     [t].[PreNotification3] * -1 / 10000

but this will not work, cause sqlserver divides the 1 before the multiplication. cause of this I changed the order of the opperands for sqlserver. I could not changes this as default, cause it does not work with native supported intervals, there the -1 needs to be at the end

So maybe our sql code generation in sqlserver is wrong, cause the expression tree I generate has the multiplication before the divission, but the sqlgenerator does not create brackets

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
implement correct timespan handling in linq2db

jogibear9988 commented:
/azp run test-all

sdanyliv commented on Source/LinqToDB/Common/Configuration.cs:
```diff
@@ -54,6 +54,11 @@ public static Delegate CompileExpression(this LambdaExpression expression)
+		/// <summary>
+		/// Fixes only partly working datediff and timespanpart calls in ExpressionBuilderSqlBuilder
+		/// </summary>
+		internal static bool DisableLegacySqlBuilderDateDiffCalls;
```
Why it is needed? Remove legacy completely.

jogibear9988 replied:
Because Legacy works if you map TimeSpan to Time Database Type. Wich is not correct but would not work with my implementation

sdanyliv commented on Source/LinqToDB/Linq/Expressions.cs:
```diff
@@ -124,6 +124,97 @@ public static void MapMember<T>(string providerName, Expression<Func<T,object?>>
 
+		#region MapUnary
+
+		static UnaryExpression GetUnaryNode(Expression expr)
```
We have `UnwrapConvert()` extension

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
use old "GetExpressionType()"

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
mysql has no millisecond

jogibear9988 committed:
fix rounding issues

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
fix firebird timespan datediff result needs to be bigint

jogibear9988 committed:
try fix clickhouse

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
one more fix for clickhouse

jogibear9988 committed:
fix spaces

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
try fix clickhouse

jogibear9988 committed:
Sybase does not work, cause of overflow

jogibear9988 commented:
/azp run test-all

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
check throwing of date functions

jogibear9988 commented:
/azp run test-all

sdanyliv committed:
Added new translators support.

jogibear9988 committed:
fix compile in lq2db 6

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
add a test for bulk copy

jogibear9988 commented:
/azp run test-all

sdanyliv commented on Source/LinqToDB/DataProvider/Access/Translation/AccessMemberTranslator.cs:
```diff
@@ -50,23 +50,29 @@ public class DateFunctionsTranslator : DateFunctionsTranslatorBase
 
 				if (partStr == null)
-					return null;
+					throw new NotImplementedException($"TranslateDateTimeDateAdd for datepart (${datepart}) not implemented");
```
Do not throw here. Return `null` or `SqlErrorExpression`.

sdanyliv commented on Source/LinqToDB/Linq/Expressions.cs:
```diff
@@ -1366,6 +1485,68 @@ public void SetExpression(LambdaExpression expression)
+
+#if NET7_0_OR_GREATER
+			MapMember<TimeSpan>(p => p.Microseconds, (Expression<Func<TimeSpan, long>>)(p => Sql.TimeSpanPart(Sql.TimeSpanParts.Microseconds, p)!.Value));
+			MapMember<TimeSpan>(p => p.TotalMicroseconds, (Expression<Func<TimeSpan, long>>)(p => Sql.TimeSpanPart(Sql.TimeSpanParts.TotalMicroseconds, p)!.Value));
+			MapMember<TimeSpan>(p => p.Nanoseconds, (Expression<Func<TimeSpan, long>>)(p => Sql.TimeSpanPart(Sql.TimeSpanParts.Nanoseconds, p)!.Value));
+			MapMember<TimeSpan>(p => p.TotalNanoseconds, (Expression<Func<TimeSpan, long>>)(p => Sql.TimeSpanPart(Sql.TimeSpanParts.TotalNanoseconds, p)!.Value));
+#endif
+
+			MapBinary((DateTime d, DateTime t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+			MapBinary((DateTime d, DateTime? t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+			MapBinary((DateTime? d, DateTime t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+			MapBinary((DateTime? d, DateTime? t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+
+			MapBinary((DateTimeOffset d, DateTimeOffset t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+			MapBinary((DateTimeOffset d, DateTimeOffset? t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+			MapBinary((DateTimeOffset? d, DateTimeOffset t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+			MapBinary((DateTimeOffset? d, DateTimeOffset? t) => d - t, (d, t) => Sql.DateDiffInterval(t, d)!.Value);
+
+			MapBinary((DateTime d, TimeSpan t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+			MapBinary((DateTime d, TimeSpan? t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+			MapBinary((DateTime? d, TimeSpan t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+			MapBinary((DateTime? d, TimeSpan? t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+
+			MapBinary((DateTimeOffset d, TimeSpan t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+			MapBinary((DateTimeOffset d, TimeSpan? t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+			MapBinary((DateTimeOffset? d, TimeSpan t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+			MapBinary((DateTimeOffset? d, TimeSpan? t) => d + t, (d, t) => Sql.DateAdd(d, t)!.Value);
+
+			MapUnary<TimeSpan, TimeSpan?>((TimeSpan t) => -t, (t) => Sql.NegateInterval(t)!.Value);
+			MapUnary<TimeSpan?, TimeSpan?>((TimeSpan? t) => -t, (t) => Sql.NegateInterval(t)!.Value);
+
+			MapBinary((DateTime d, TimeSpan t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
+			MapBinary((DateTime d, TimeSpan? t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
+			MapBinary((DateTime? d, TimeSpan t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
+			MapBinary((DateTime? d, TimeSpan? t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
+
+			MapBinary((DateTimeOffset d, TimeSpan t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
+			MapBinary((DateTimeOffset d, TimeSpan? t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
+			MapBinary((DateTimeOffset? d, TimeSpan t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
+			MapBinary((DateTimeOffset? d, TimeSpan? t) => d - t, (d, t) => Sql.DateAdd(d, -t)!.Value);
```
Do not extend `Expressions.cs`, use MemberTranslator.

jogibear9988 replied:
I don't get how to do this in 
ProviderMemberTranslatorDefault or MemberTranslatorBase
can you give me a sample?

There is nearly the same code for all other Types in Expression.s, why should this not be there?

sdanyliv replied:
I have added IBinaryTranslator and IUnaryTranslator

sdanyliv commented on Source/LinqToDB/SqlQuery/QueryHelper.cs:
```diff
@@ -343,6 +343,21 @@ static bool IsDependsOn(IQueryElement testedRoot, IQueryElement onElement, HashS
 			return null;
 		}
 
+		public static DbDataType GetExpressionType(this ISqlExpression expr)
```
Looks like duplicate of `GetDbDataType`

jogibear9988 replied:
Need to check if it is still needed, it's now 2 years since this is open. I'll look

sdanyliv reviewed (commented):
A little outdated implementation.

jogibear9988 commented on Source/LinqToDB/DataProvider/Access/Translation/AccessMemberTranslator.cs:
```diff
@@ -36,7 +36,7 @@ public class DateFunctionsTranslator : DateFunctionsTranslatorBase
 
 				if (partStr == null)
-					return null;
+					throw new NotImplementedException($"TranslateDateTimeDatePart for datepart (${datepart}) not implemented");
```
<img width="1290" alt="image" src="https://github.com/user-attachments/assets/d8f377c9-dd51-49ca-8cd8-4f23604725c4" />
 SQLError expression seems not to work....
 
 But is null better? Then I don't get what the error is

jogibear9988 commented:
> A little outdated implementation.

It's open for a long time ;-)

What is outdated?

sdanyliv commented:
What exactly is considered outdated?

These translations should be implemented using `MemberTranslator`. It's a more robust and flexible approach, allowing us to handle a wider range of scenarios and database-specific variations.

jogibear9988 commented:
@sdanyliv 

Can you help me, how to implement this in MemberTranslator? And how to do it so it works for all databases?

But before, I tried my code, and my MapBinary & MapUnary calls I've made, seem not to be called any more. Was there a change wich broke these? I think they still should work, even if we change to MemberTranslator afterwards.

sdanyliv committed:
Merge branch 'master' into feature/4726

sdanyliv committed:
Merge branch 'feature/4726' into issue/3993

sdanyliv commented:
>  MapBinary & MapUnary

Probably new translator do not care about this. I have added to this PR `IBinaryTranslator` and `IUnaryTranslator`. Can you look at this?

sdanyliv commented:
Please let me know which test you believe *should* be working but currently isn't.

jogibear9988 committed:
Merge branch 'master' of https://github.com/linq2db/linq2db into issue/3993new

# Conflicts:
#	Source/LinqToDB/Common/Configuration.cs
#	Source/LinqToDB/Internal/DataProvider/SqlServer/Translation/SqlServerMemberTranslator.cs
#	Source/LinqToDB/Internal/DataProvider/Translation/SqlExpressionFactoryExtensions.cs
#	Source/LinqToDB/Sql/Sql.DateTime.cs

jogibear9988 committed:
fix compilation after merge of master

jogibear9988 committed:
little test code cleanup

jogibear9988 commented:
> Please let me know which test you believe _should_ be working but currently isn't.

I map here the Binary expressions: (a datatime minus a timespan)
https://github.com/linq2db/linq2db/pull/3994/files#diff-2a6a9163861fbdc4015c6f32a3373c85469cb401552aea366b2d2c07ede63fb5R1557

So when I run this test and break here: (for example for postgres)
https://github.com/linq2db/linq2db/pull/3994/files#diff-71d51f038daa00c195480c91854cacefd7ece6c3e109b5b48d7c99fb3d9e77adR104

Then I should have landed in DateTimeAddIntervalBuilderPostgreSQL:
https://github.com/linq2db/linq2db/pull/3994/files#diff-0381ba7e238d0f1809d47075f54b06e6c5fb0dfd9c7f0ed1017188c0eab15e38R277

but this is not called any more.

So the MapBinary does not seem to have any effect, or some other cancels this out....

jogibear9988 committed:
cleanup some code and tests

jogibear9988 committed:
hack to add back mapBinary support

jogibear9988 commented:
/azp run test-all

jogibear9988 committed:
fix coding style errors

jogibear9988 commented:
/azp run test-all

sdanyliv committed:
Fix formatting.

sdanyliv committed:
Revert hack.

sdanyliv committed:
Translation fix.

sdanyliv committed:
Fix build.

sdanyliv commented:
@jogibear9988, I need to rethink this PR. Something wrong in translation decision, espceially hoe to handle timespan in the SQL. Will take a look tomorrow.

sdanyliv commented:
/azp run test-all

jogibear9988 commented:
> @jogibear9988, I need to rethink this PR. Something wrong in translation decision, espceially hoe to handle timespan in the SQL. Will take a look tomorrow.

okay....

Test4 can be removed... It does not work for sqlite cause of missing Mapping. But I don't think it is needed, I tested there a issue with DateTime part access wich did not work after I refcatored something.

sdanyliv committed:
Merge branch 'master' into issue/3993

jogibear9988 committed:
Merge branch 'master' of https://github.com/linq2db/linq2db into issue/3993new

jogibear9988 committed:
fix compilation with visual studio 2026

jogibear9988 commented:
@copilot can you review?

jogibear9988 commented:
replaced by #5334
